### ISSUE TITLE
Nested merge-joins query causes NoMethodError with ActiveRecord 6.1.4.4

### ISSUE THREAD

**Issue by yysaki** [rzane/baby_squeel#119]
#### Issue

Like as #117, this problem occurs with ActiveRecord 6.1.4.4 and later, but doesn't occur with ActiveRecord 6.0.4.4 and below.
The following nested merge-joins query causes NoMethodError.

```
`method_missing': private method `construct_join_dependency' called for #<Post::ActiveRecord_Relation:0x00007f93d9fa2cc8> (NoMethodError)
```

#### Reproduction

```ruby
require 'bundler/inline'
require 'minitest/spec'
require 'minitest/autorun'

gemfile true do
  source 'https://rubygems.org'
  gem 'activerecord', '~> 6.1.4.4' # which Active Record version?
  gem 'sqlite3'
  gem 'baby_squeel'
end

ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')

ActiveRecord::Schema.define do
  create_table :authors, force: true do |t|
    t.string :name
  end
  create_table :posts, force: true do |t|
    t.belongs_to :author
    t.string :title
  end
  create_table :comments, force: true do |t|
    t.belongs_to :post
    t.string :body
  end
end

class Author < ActiveRecord::Base
  has_many :posts
end

class Post < ActiveRecord::Base
  belongs_to :author
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BabySqueelTest < Minitest::Spec
  it 'works' do
    scope = Author.joins(:posts).merge(Post.joins(:comments).merge(Comment.where(body: 'body')))
    scope.to_sql.must_equal %{
      SELECT "authors".* FROM "authors"
      INNER JOIN "posts" ON "posts"."author_id" = "authors"."id"
      INNER JOIN "comments" ON "comments"."post_id" = "posts"."id"
      WHERE "comments"."body" = 'body'
    }.squish
  end
end
```

<details>
<summary>Test result</summary>

```
Run options: --seed 5388

# Running:

Traceback (most recent call last):
        26: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:73:in `block in autorun'
        25: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:146:in `run'
        24: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:169:in `__run'
        23: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:169:in `map'
        22: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:169:in `block in __run'
        21: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:335:in `run'
        20: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:363:in `with_info_handler'
        19: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:376:in `on_signal'
        18: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:336:in `block in run'
        17: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:336:in `each'
        16: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:337:in `block (2 levels) in run'
        15: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:350:in `run_one_method'
        14: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:1042:in `run_one_method'
        13: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:93:in `run'
        12: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:221:in `with_info_handler'
        11: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:376:in `on_signal'
        10: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:94:in `block in run'
         9: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:281:in `time_it'
         8: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:95:in `block (2 levels) in run'
         7: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:195:in `capture_exceptions'
         6: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:98:in `block (3 levels) in run'
         5: from issue.rb:43:in `block in <class:BabySqueelTest>'
         4: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/activerecord-6.1.4.4/lib/active_record/relation/spawn_methods.rb:35:in `merge'
         3: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/activerecord-6.1.4.4/lib/active_record/relation/spawn_methods.rb:46:in `merge!'
         2: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/activerecord-6.1.4.4/lib/active_record/relation/merger.rb:82:in `merge'
         1: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/activerecord-6.1.4.4/lib/active_record/relation/merger.rb:122:in `merge_joins'
/Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/activerecord-6.1.4.4/lib/active_record/relation/delegation.rb:110:in `method_missing': private method `construct_join_dependency' called for #<Post::ActiveRecord_Relation:0x00007f93d9fa2cc8> (NoMethodError)
        22: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:73:in `block in autorun'
        21: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:146:in `run'
        20: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:169:in `__run'
        19: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:169:in `map'
        18: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:169:in `block in __run'
        17: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:335:in `run'
        16: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:363:in `with_info_handler'
        15: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:376:in `on_signal'
        14: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:336:in `block in run'
        13: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:336:in `each'
        12: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:337:in `block (2 levels) in run'
        11: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:350:in `run_one_method'
        10: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:1042:in `run_one_method'
         9: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:93:in `run'
         8: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:221:in `with_info_handler'
         7: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:376:in `on_signal'
         6: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:94:in `block in run'
         5: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest.rb:281:in `time_it'
         4: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:95:in `block (2 levels) in run'
         3: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:194:in `capture_exceptions'
         2: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:201:in `rescue in capture_exceptions'
         1: from /Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:205:in `sanitize_exception'
/Users/yysaki/.anyenv/envs/rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/minitest-5.15.0/lib/minitest/test.rb:205:in `dump': instance variable added to NoMethodError instance (RuntimeError)
```
</details>

**Comment by rocket-turtle** [comment]
Thank you for your great bug report @yysaki.

I'll work on this on monday but I think the fix should be easy. The `construct_join_dependency` is not private.

**Comment by rocket-turtle** [comment]
@rzane this fix was quit small. I didn't pay attention to the method visibility :/

@yysaki Sry for your bad upgrade experiance. I got the baby_squeel tests green but haven't used the version in our application because we are still upgrading to Rails 6.0

Your test is green wirh:
`gem 'baby_squeel', branch: 'ISSUE-119', git: 'https://github.com/rocket-turtle/baby_squeel'`

If you find any other bugs please feel free to report them.

**Comment by rzane** [comment]
Fixed in #120
